source/Background/Controller.js

```diff
@@ -323,72 +324,59 @@ backgroundController.exposeController(
-        // && !!canisterInfo.category && canisterInfo.category in PROTECTED_CATEGORIES);
-
-        if (shouldShowModal) {
-          const url = qs.stringifyUrl({
-            url: 'notification.html',
-            query: {
-              callId,
-              portId,
-              type: 'sign',
-              metadataJson: JSON.stringify(metadata),
-              argsJson: JSON.stringify({
-                requestInfo,
-                payload,
-                canisterInfo,
-                timeout: app?.timeout,
-              }),
-            },
-          });
-
-          const height = keyring?.isUnlocked
-            ? SIZES.appConnectHeight
-            : SIZES.loginHeight;
-
-          extension.windows.create({
-            url,
-            type: 'popup',
-            width: SIZES.width,
-            height,
+      const isDangerousUpdateCall = !preApprove && requestType === 'call';
+      if (isDangerousUpdateCall) {
+        getApps(keyring.currentWalletId.toString(), async (apps = {}) => {
+          const app = apps?.[metadata.url] || {};
+          if (app.status !== CONNECTION_STATUS.accepted) {
+            callback(ERRORS.CONNECTION_ERROR, null);
+            return;
+          }
+          if (canisterId && !(canisterId in app.whitelist)) {
+            callback(ERRORS.CANISTER_NOT_WHITLESTED_ERROR(canisterId), null);
+            return;
+          }
+          getProtectedIds(async (protectedIds) => {
+            const canisterInfo = app.whitelist[canisterId];
+            const shouldShowModal = protectedIds.includes(canisterInfo.id);
+
+            if (shouldShowModal) {
+              const url = qs.stringifyUrl({
+                url: 'notification.html',
+                query: {
+                  callId,
+                  portId,
+                  type: 'sign',
+                  metadataJson: JSON.stringify(metadata),
+                  argsJson: JSON.stringify({
+                    requestInfo,
+                    payload,
+                    canisterInfo,
+                    timeout: app?.timeout,
+                  }),
+                },
+              });
+              const height = keyring?.isUnlocked
+                ? SIZES.appConnectHeight
+                : SIZES.loginHeight;
+              extension.windows.create({
+                url,
+                type: 'popup',
+                width: SIZES.width,
+                height,
+              });
+            } else {
+              const parsedPayload = new Uint8Array(Object.values(payload));
+              const signed = await keyring.sign(parsedPayload.buffer);
+              callback(null, [...new Uint8Array(signed)]);
+            }
           });
-        } else {
-          const parsedPayload = new Uint8Array(Object.values(payload));
-
-          const signed = await keyring.sign(parsedPayload.buffer);
-          callback(null, [...new Uint8Array(signed)]);
-        }
-      });
+        });
+      } else {
```

Comment:
We could reverse the `if` conditional and return the method, that way we avoid one extra level of `if` nesting, (not really required)

Reply:
Cannot do that due to performance. To do that I'd need to get all protected IDs from storage on every signature call instead of only when needed (when it's a dangerous call)